File: web/js/core/contract.js
// 계약 / 연봉 / 서비스타임 / 보유권 / 구단 재정. 단위는 억(원).
export const MIN_SALARY = 0.30, WAR_PRICE = 4.2, FA_SERVICE = 7, ARB_SERVICE = 3;

export class Contract {
  constructor(startYear, salaries) { this.start_year = startYear; this.salaries = [...salaries]; }
  get end_year() { return this.start_year + this.salaries.length - 1; }
  get years() { return this.salaries.length; }
  get total() { return this.salaries.reduce((a,b)=>a+b, 0); }
  get aav() { return this.total / this.years; }
  salaryIn(y) { const i = y - this.start_year; return (i>=0 && i<this.salaries.length) ? this.salaries[i] : 0; }
  remaining(y) { return this.salaries.slice(Math.max(0, y - this.start_year)); }
  toString() { return `${this.years}년 ${this.total.toFixed(1)}억`; }
}

export function projWar(ovr, age, isPitcher, ahead = 0) {
  const a = age + ahead;
  let o = ovr;
  if (a < 27 && ahead) o += Math.min(2.5, 0.9*(27-age));
  if (a > 29) o -= 0.85*(a-29);
  return Math.max(-0.8, (o - 42) * (isPitcher ? 0.19 : 0.22));
}
export function marketValue(ovr, age, isPitcher) {
  return Math.max(MIN_SALARY, projWar(ovr, age, isPitcher) * WAR_PRICE);
}
export function demandYears(age, ovr) {
  if (ovr < 46) return 1;
  if (age <= 27) return ovr >= 55 ? 5 : 3;
  if (age <= 30) return ovr >= 58 ? 4 : 3;
  if (age <= 33) return ovr >= 58 ? 3 : 2;
  return ovr >= 60 ? 2 : 1;
}
/** 보유권까지 포함한 팀 통제 기간의 연도별 연봉. */
export function controlHorizon(p, ovr, year, isPitcher, maxYears = 6) {
  const rem = p.contract ? [...p.contract.remaining(year+1)] : [];
  const sv = (p.service ?? 0) + rem.length;
  const extra = Math.max(0, Math.min(maxYears - rem.length, FA_SERVICE - sv));
  for (let i = 0; i < extra; i++) {
    const age = p.age + rem.length + 1;
    rem.push(Math.max(MIN_SALARY, marketValue(ovr, age, isPitcher) * Math.min(0.9, 0.32 + 0.16*i)));
  }
  return rem.slice(0, maxYears);
}
export function renewalSalary(p, ovr, isPitcher) {
  const sv = p.service ?? 0;
  const mv = marketValue(ovr, p.age, isPitcher);
  const frac = sv < ARB_SERVICE ? 0.04 + 0.045*sv : 0.32 + 0.16*(sv - ARB_SERVICE);
  return Math.max(MIN_SALARY, Math.round(mv * Math.min(0.92, frac) * 100) / 100);
}
export function isFreeAgent(p, year) {
  if ((p.service ?? 0) < FA_SERVICE) return false;
  return !p.contract || p.contract.end_year <= year;
}

// 관중 · 수입 단가 (억 원)
export const TICKET = 0.00006;   // 1인당 입장 수입 ≈ 6,000원
export const CONCESSION = 0.00003; // 1인당 식음료·굿즈 ≈ 3,000원

/** 한 경기 관중 동원율. 성적과 시장 규모가 팬을 부른다. */
export function attendRate(f, winPct, playoffLast, titleLast, rng) {
  let r = 0.55
    + 0.22 * (f.market_size - 1)
    + 0.22 * (winPct - 0.5)
    + (playoffLast ? 0.10 : 0)
    + (titleLast ? 0.08 : 0)
    + (rng ? rng.gauss(0, 0.07) : 0);
  return Math.max(0.15, Math.min(1.0, r));
}

export class Finance {
  /** @param fr 연고 구단의 성격. 없으면 리그 평균으로 뽑는다.
   *  시장 규모를 매 판 난수로 뽑으면 부산이 어떤 게임에서는 작은 시장이 된다.
   *  구단에 이야기가 붙으려면 이 값이 구단을 따라다녀야 한다. */
  constructor(rng, fr = null) {
    this.market_size = Math.max(0.55, Math.min(1.5,
      rng.gauss(fr && fr.market ? fr.market : 1.0, fr ? 0.07 : 0.24)));
    this.owner_spending = Math.max(0.6, Math.min(1.4, rng.gauss(1.0, 0.16)));
    // 구단주 인내심 — 큰 시장일수록 짧다. 성적이 곧 압박이 된다.
    this.patience = Math.max(15, Math.min(85,
      rng.gauss(52 - (this.market_size - 1) * 26 + (fr ? fr.temper || 0 : 0), 11)));
    this.revenue = 100 * this.market_size;
    this.budget = 100 * this.market_size;
    this.attendance = 0;      // 지난 시즌 홈 총관중
    this.homeGames = 0;
    this.income = null;       // 수입 구성
  }

  /** 구단주가 이번 시즌에 요구하는 것. 시장 규모와 인내심이 정한다. */
  get demand() {
    const p = this.market_size * 1.6 - this.patience / 55;
    if (p >= 1.35) return '우승';
    if (p >= 0.85) return '포스트시즌';
    if (p >= 0.35) return '5할 승률';
    return '재건 허용';
  }
  /** 관중이 수입을 만들고, 수입이 다음 시즌 예산이 된다. */
  update(winPct, playoffs, title, attendance = 0, homeGames = 0, capacity = 18000) {
    if (!attendance) {                      // 기록이 없으면 추정치로 채운다
      homeGames = homeGames || 72;
      attendance = capacity * attendRate(this, winPct, playoffs, title) * homeGames;
    }
    this.attendance = Math.round(attendance);
    this.homeGames = homeGames;
    const ticket = attendance * TICKET;
    const conc = attendance * CONCESSION;
    // 중계권은 리그가 균등 배분한다. 이게 없으면 큰 시장이 눈덩이처럼 커진다.
    const media = 50 + 9 * this.market_size + (playoffs ? 6 : 0) + (title ? 4 : 0);
    this.income = { ticket: +ticket.toFixed(1), concession: +conc.toFixed(1),
                    media: +media.toFixed(1) };
    this.revenue = ticket + conc + media;
    // 예산은 관성이 있다. 한 시즌 성적으로 갑자기 두 배가 되지 않는다.
    const target = this.revenue * 0.80 * this.owner_spending;
    this.budget = this.budget ? this.budget * 0.55 + target * 0.45 : target;
  }
}
/** 경쟁균형세 — KBO 에도 있는 제동장치.
 *  성적→관중→수입→전력의 되먹임에 브레이크가 없으면 한 팀이 리그를 독식한다. */
export function balanceBudgets(teams) {
  const mean = teams.reduce((s, t) => s + t.finance.budget, 0) / teams.length;
  for (const t of teams) {
    const f = t.finance;
    if (f.budget > mean * 1.28) f.budget = mean * 1.28 + (f.budget - mean * 1.28) * 0.35;
    if (f.budget < mean * 0.74) f.budget = mean * 0.74 - (mean * 0.74 - f.budget) * 0.35;
  }
}

export const payroll = (t, year) =>
  [...t.batters, ...t.pitchers].reduce((s,p) => s + (p.contract ? p.contract.salaryIn(year) : 0), 0);

File: proto/contract.py
"""
Project Dugout - Contract & Finance

단위는 '억'(원). KBO 감각에 맞췄다.
  최저연봉 0.3억 / 팀 연봉총액 80~160억 / 최상급 FA 연 20억대
"""
import development as dev
from pa_engine import Batter

MIN_SALARY = 0.30
WAR_PRICE = 4.2          # FA 시장에서 1 WAR 의 가격 (억)
FA_SERVICE = 7           # FA 자격 취득에 필요한 1군 등록 시즌
ARB_SERVICE = 3          # 연봉조정 자격


class Contract:
    def __init__(self, start_year, salaries, signed_by=None):
        self.start_year = start_year
        self.salaries = list(salaries)       # 연차별 연봉
        self.signed_by = signed_by

    @property
    def end_year(self):
        return self.start_year + len(self.salaries) - 1

    @property
    def years(self):
        return len(self.salaries)

    @property
    def total(self):
        return sum(self.salaries)

    @property
    def aav(self):
        return self.total / self.years

    def salary_in(self, year):
        i = year - self.start_year
        return self.salaries[i] if 0 <= i < len(self.salaries) else 0.0

    def remaining(self, year):
        i = max(0, year - self.start_year)
        return self.salaries[i:]

    def __str__(self):
        return f"{self.years}년 {self.total:.1f}억"


# ---------------------------------------------------------------------------
# 가치 평가
# ---------------------------------------------------------------------------
def proj_war(ovr, age, is_pitcher, seasons_ahead=0):
    """능력치 → 예상 시즌 WAR. 앞으로의 노화까지 반영한다."""
    a = age + seasons_ahead
    o = ovr
    if a < 27:
        o += min(2.5, 0.9 * (27 - age)) * (1 if seasons_ahead else 0)
    if a > 29:
        o -= 0.85 * (a - 29)
    k = 0.19 if is_pitcher else 0.22
    return max(-0.8, (o - 42.0) * k)


def market_value(ovr, age, is_pitcher):
    """FA 시장에서의 연평균 몸값."""
    w = proj_war(ovr, age, is_pitcher)
    return max(MIN_SALARY, w * WAR_PRICE)


def demand_years(age, ovr):
    """선수가 원하는 계약 기간."""
    if ovr < 46:
        return 1
    if age <= 27:
        return 5 if ovr >= 55 else 3
    if age <= 30:
        return 4 if ovr >= 58 else 3
    if age <= 33:
        return 3 if ovr >= 58 else 2
    return 2 if ovr >= 60 else 1


def control_horizon(p, ovr, year, is_pitcher, max_years=6):
    """팀이 이 선수를 보유하는 기간의 연도별 예상 연봉.

    계약 잔여 연수로만 계산하면 안 된다. 계약이 끝나도 FA 자격 전이면
    보류권이 남아 값싸게 계속 쓸 수 있다. 이 보유 기간이 곧 자산의 크기다.
    (이걸 빼먹으면 '1년 계약 24세'가 리빌딩 팀에게 무가치해진다.)
    """
    c = p.contract
    rem = list(c.remaining(year + 1)) if c else []
    sv = getattr(p, "service", 0) + len(rem)
    for i in range(max(0, min(max_years - len(rem), FA_SERVICE - sv))):
        age = p.age + len(rem) + 1
        frac = 0.32 + 0.16 * i
        rem.append(max(MIN_SALARY, market_value(ovr, age, is_pitcher) * min(0.9, frac)))
    return rem[:max_years]


def surplus_value(p, ovr, year, is_pitcher):
    """잉여 가치 = 앞으로의 생산 - 앞으로의 연봉. 트레이드 가치의 뼈대."""
    c = p.contract
    rem = c.remaining(year + 1) if c else []
    # 계약이 끝나면 그 뒤는 우리 것이 아니다. 단 FA 자격 전이면 보유권이 남는다
    if not rem:
        yrs_left = max(0, FA_SERVICE - getattr(p, "service", 0))
        rem = [market_value(ovr, p.age, is_pitcher) * 0.45] * min(3, yrs_left)
    s = 0.0
    for i, sal in enumerate(rem):
        w = proj_war(ovr, p.age, is_pitcher, seasons_ahead=i + 1)
        s += w * WAR_PRICE - sal
    return s


# ---------------------------------------------------------------------------
# 구단 재정
# ---------------------------------------------------------------------------
class Finance:
    def __init__(self, rng):
        self.market_size = max(0.55, min(1.5, rng.gauss(1.0, 0.24)))
        self.owner_spending = max(0.6, min(1.4, rng.gauss(1.0, 0.16)))
        self.revenue = 100.0 * self.market_size
        self.budget = 100.0 * self.market_size

    def update(self, win_pct, made_playoffs, won_title):
        base = 62.0 + 58.0 * self.market_size
        base *= 0.80 + 0.62 * (win_pct / 0.5) * 0.5      # 성적이 수입을 좌우
        if made_playoffs:
            base += 9.0 * self.market_size
        if won_title:
            base += 7.0 * self.market_size
        self.revenue = base
        self.budget = base * 0.86 * self.owner_spending


def payroll(t, year):
    return sum(p.contract.salary_in(year) for p in t.batters + t.pitchers
               if p.contract)


# ---------------------------------------------------------------------------
# 계약 갱신 (FA 자격 전)
# ---------------------------------------------------------------------------
def renewal_salary(p, ovr, is_pitcher):
    """FA 자격 전 선수의 연봉. 서비스 타임에 따라 시장가의 일부만 받는다."""
    sv = getattr(p, "service", 0)
    mv = market_value(ovr, p.age, is_pitcher)
    if sv < ARB_SERVICE:
        frac = 0.04 + 0.045 * sv
    else:
        frac = 0.32 + 0.16 * (sv - ARB_SERVICE)
    return max(MIN_SALARY, round(mv * min(0.92, frac), 2))


def is_free_agent(p, year):
    """계약이 끝났고 FA 자격을 채웠으면 FA."""
    if getattr(p, "service", 0) < FA_SERVICE:
        return False
    return (p.contract is None) or (p.contract.end_year <= year)
